### PR TITLE
Record failed authentication attempts

### PR DESCRIPTION
**Pull request by garyhtou**

### CODE REVIEW

File: app/controllers/users/sessions_controller.rb [hunk 1]
```diff
@@ -4,7 +4,20 @@ class Users::SessionsController < ApplicationController
...
+
+      if authentication.succeeded?
+        authentication.reject reason: :previously_succeeded
+        return redirect_to sign_in_path
+      end
+
+      unless authentication.user.admin? # Don't allow non-admins to sign in
+        authentication.reject reason: :non_admin
+        return redirect_to root_path
+      end
```
Comment (northeastprince) [inline app/controllers/users/sessions_controller.rb:20]: This might be a bit confusing for someone who's not an admin and doesn't know what's going on.

**Review by northeastprince — COMMENTED** [review]
Love that we're keeping track of failed authentications!

File: app/controllers/users/sessions_controller.rb (hunk 1 above)
Comment (garyhtou) [reply app/controllers/users/sessions_controller.rb]: I also just remembered that https://github.com/hackclub/hackathons-backend/pull/77 added `Hackathons::Submissions#index` which allows signed-in non-admins to view their applications

File: test/system/authentication_test.rb [hunk 2]
```diff
@@ -46,6 +46,8 @@ class AuthenticationTest < ApplicationSystemTestCase
     end
 
     assert_not authentication.succeeded?
+    assert_equal "rejected", authentication.events&.last&.action
+    assert_equal "expired", authentication.events&.last&.details&.dig("reason")
```
Comment (northeastprince) [inline test/system/authentication_test.rb:50]: Might wanna add these assertions for the invalid link test too.

**Review by northeastprince — APPROVED** [review]
LGTM!

File: test/system/authentication_test.rb (hunk 2 above)
Comment (garyhtou) [reply test/system/authentication_test.rb:50]: With an invalid link, we aren't able to match the request to an Authentication record. So, there's no Authentication for us to record Events on to

**Pull request merged**